File: apps/cron/src/jobs/keys.ts
import type { Database } from "@raven/db";
import { virtualKeys } from "@raven/db";
import { and, eq, isNotNull, lt } from "drizzle-orm";

export const deactivateExpiredKeys = async (db: Database) => {
  const now = new Date();

  const deactivated = await db
    .update(virtualKeys)
    .set({ isActive: false })
    .where(
      and(
        isNotNull(virtualKeys.expiresAt),
        lt(virtualKeys.expiresAt, now),
        eq(virtualKeys.isActive, true)
      )
    )
    .returning({ id: virtualKeys.id });

  console.log(
    `Key deactivation: deactivated ${deactivated.length} expired keys`
  );
};

File: apps/cron-py/src/jobs/keys.py
def deactivate_expired_keys(conn) -> None:
    with conn.cursor() as cur:
        cur.execute("""
            UPDATE virtual_keys
            SET "isActive" = false
            WHERE "isActive" = true
              AND "expiresAt" IS NOT NULL
              AND "expiresAt" < NOW()
        """)
        count = cur.rowcount
        if count > 0:
            print(f"  Deactivated {count} expired key(s)")
